external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java

```diff
@@ -161,7 +176,11 @@ public void prepareCommit(long txid) {
             commands = jedisContainer.getInstance();
             if (commands.exists(prepareNamespace)) {
                 LOG.debug("Prepared txn already exists, will merge", txid);
-                pendingPrepare.putAll(pendingCommit);
+                for (Map.Entry<String, String> e: pendingCommit.entrySet()) {
+                    if (!pendingPrepare.containsKey(e.getKey())) {
+                        pendingPrepare.put(e.getKey(), e.getValue());
+                    }
+                }
             }
             if (!pendingPrepare.isEmpty()) {
                 commands.hmset(prepareNamespace, pendingPrepare);
```

Comment:
A potential issue (already existing) which could be fixed here or in a separate JIRA. If the pendingPrepare is updated after the hmset but before its assigned to pendingCommit, the pendingCommit will not reflect the entities in prepareNamespace. I think if the pendingPrepare is snapshotted/copied before hmset and the copy is used for both hmset and assigned to pendingCommit it would avoid this race condition.